File: 2018-minimum-space-wasted-from-packaging/minimum-space-wasted-from-packaging.rb
def min_wasted_space(packages, boxes)
  mod = 1_000_000_007

  packages.sort!

  n = packages.length

  prefix = Array.new(n + 1, 0)
  (0...n).each do |i|
    prefix[i + 1] = prefix[i] + packages[i]
  end

  ans = Float::INFINITY

  boxes.each do |supplier|
    supplier.sort!

    next if supplier[-1] < packages[-1]

    waste = 0
    idx = 0

    supplier.each do |b|
      nxt = upper_bound(packages, idx, b)

      waste += (nxt - idx) * b - (prefix[nxt] - prefix[idx])

      idx = nxt
      break if idx == n
    end

    ans = [ans, waste].min
  end

  ans == Float::INFINITY ? -1 : ans % mod
end

def upper_bound(arr, start, target)
  l = start
  r = arr.length

  while l < r
    mid = (l + r) / 2
    if arr[mid] <= target
      l = mid + 1
    else
      r = mid
    end
  end

  l
end

File: 2018-minimum-space-wasted-from-packaging/minimum-space-wasted-from-packaging.rs
impl Solution {
    pub fn min_wasted_space(mut packages: Vec<i32>, boxes: Vec<Vec<i32>>) -> i32 {
        const MOD: i64 = 1_000_000_007;

        packages.sort();

        let n = packages.len();

        let mut prefix = vec![0i64; n + 1];
        for i in 0..n {
            prefix[i + 1] = prefix[i] + packages[i] as i64;
        }

        let mut ans = i64::MAX;

        for mut supplier in boxes {
            supplier.sort();

            if *supplier.last().unwrap() < packages[n - 1] {
                continue;
            }

            let mut waste = 0i64;
            let mut idx = 0usize;

            for &b in &supplier {
                let next = upper_bound(&packages, idx, b);

                waste += (next - idx) as i64 * b as i64 - (prefix[next] - prefix[idx]);

                idx = next;
                if idx == n {
                    break;
                }
            }

            ans = ans.min(waste);
        }

        if ans == i64::MAX {
            -1
        } else {
            (ans % MOD) as i32
        }
    }
}

fn upper_bound(arr: &Vec<i32>, start: usize, target: i32) -> usize {
    let mut l = start;
    let mut r = arr.len();

    while l < r {
        let mid = (l + r) / 2;
        if arr[mid] <= target {
            l = mid + 1;
        } else {
            r = mid;
        }
    }

    l
}